### DISCUSSION TITLE
VScode setup

### DISCUSSION THREAD

**Discussion by MadTooler** [grblHAL/RP2040#40] [General]
I see that the RP2040 grblHAL project is noted as a VScode project.  I keep looking for some directions to setup the grblHAL/RP2040 project in VScode (free version) so I can make my own pin and machine settings, but I have not had any luck.  I am new to VScode with only a PlatformIO setup for programming other projects on the RP2040, so I am probably overlooking the obvious.  However, I have also tried to import into the Arduino IDE without any luck when I try to apply the import .ZIP suggestions noted.

I saw some instructions for PlatformIO with the core grblHal, but again I fell short of success trying to convert those instructions to the pico version.  

In the end, what I need is to be able to tweak the pin and machine settings for a breadboard version with the pico before I move on to implementation of the pico on my custom control board.  Any pointers to existing instructions and/or the simplest approach to basic config settings?

Thanks!

**Comment by terjeio** [comment]
>  I keep looking for some directions to setup the grblHAL/RP2040 project in VScode (free version)

Which OS?

> but I have not had any luck.

What is going wrong? Setting up VSCode or loading the project?

>  However, I have also tried to import into the Arduino IDE without any luck when I try to apply the import .ZIP suggestions noted.

I am using the RP2040 SDK for the driver, not the Arduino framework. I do not know if it is possible to support the Arduino framework alongside the SDK code and how much work it would be to get that working.

> I saw some instructions for PlatformIO with the core grblHal, but again I fell short of success trying to convert those instructions to the pico version.

PlatformIO does not support the Pico SDK so cannot be used.
FYI for the [Web Builder](http://svn.io-engineering.com:8080/) I use a [docker instance](https://github.com/palexander23/RP2040-Docker) for compiling.

>  Any pointers to existing instructions and/or the simplest approach to basic config settings?

There are instructions on the internet for setting up VScode for different OSes. I am using a Raspberry Pi 400 for development - setting that up was simple, basically running the RPi installation script.

**Reply by MadTooler** [reply]
Thanks for the reply and notes.

Currently running Windows 10 64b.

I have VSCode setup with platformIO, C++ extensions, and the pico board platform.  I can program and send to the pico my programs with PIO scripts and arduino style programming.  I assumed the RP2040 SDK is in that mix since it is working at that level.  

I am not clear on how to bring/load the grblHAL project into the VSCode setup.  From the instructions on github in the grblHAL repositories, I find some instruction on the other variants, but not for RP2040.  When I try to translate them, I am either missing something or doing something wrong.  Looking over other tutorials on the web with the same end results.  I get the feeling I am mashing up non compatible configurations.  Maybe I can't have platformIO and the RP2040 SDK based projects in the same install??  If there is a suitable tutorial that you find most in line, please point me at it.

My end goal is an rpi 4 GUI setup with several RP2040 picos handling various machine tasks, one being grblHAL, as a multiprocessor CNC motion controller.  The software and most hardware will be open source.  Maybe having the VSCcode setup on the rpi 4 would be a good way to have the whole system easier for users to get an all in one setup up and running.  I wonder if that would make for too slow of development environment if I asked the rpi 4 to do all of that.  

For the RPI installation script, are you referring to page 4 of https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf
Beyond that, what else is needed to load the grlblHAL project?

**Reply by MadTooler** [reply]
Here are the basic steps I took to load the project on a RPI 4b running the 64bit OS.  Some of the popups steps I did are only sort of accurate on the specifics since I didn't take notes as I accepted them.

- I ran the RPI installation script on an rpi 4b that is only on page 4 of the datasheet noted in my prior reply.  
- I then opened VSCode.
- From the terminal within VSCode I ran "git clone --recurse-submodules https://github.com/grblHAL/RP2040.git"
- Then File>Open Folder and opened the RP2040 folder under the pi home directory
- After it loaded, eventually a popup at the bottom right of VSCode said something about let CMake setup the project.  I said yes.
- VSCode popped up another notice about loading MS C++ tools and I said yes again.
- I tested the build process by clicking on the build button at bottom of screen.  It gave a few options for which compiler or to let CMake determine.  I told it to use CMake.  It said something about not being defined and would use most common. During the build, it did succeed, but I noticed a couple of things not being included or No Install step for.  Not sure if that was normal for the default configuration or a result of the compiler that was used.
- After the build, I found the build directory within RP2040.  It did generate a "grblHAL.uf2" file.  I manually copied that to my pico.  I need to read up on if that is the best method or if directly sending it from within VSCode is better (like I do with platformIO).
For reference, once I realized all I need to do to load the project is open the folder from within VSCode, I tried that from my windows machine as well.  It did load but failed to build properly.  That was somewhat of a test to see if my prior installations to use platformIO installed the RP2040 SDK in the background.  I assume if I follow the rest of the instructions in the Getting Started datasheet for pico, it should work OK.

Now I am on to figuring out which RPI compatible sender works best with grblHAL and decide my pin and machine settings.

**Comment by MadTooler** [comment]
@terjeio 
I have cloned the most recent and now I cannot get it to compile.  Which compiler should be used with the most current repository?

**Reply by MadTooler** [reply]
I cannot get it to compile without error.  Did I miss something new that needs to be also installed since my prior clone about 6 months ago?

**Reply by terjeio** [reply]
> Which compiler should be used with the most current repository?

I build with the compiler that came with the default SDK installation (on a RP400) - I did not update this when switching SDK version. See para 2.2 in the [documentation](https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf).

> Did I miss something new that needs to be also installed since my prior clone about 6 months ago?

SDK v1.5.0 as documented in the [readme](https://github.com/grblHAL/RP2040/blob/master/README.md).

**Reply by MadTooler** [reply]
```
[build] gmake[2]: *** [pico-sdk/src/rp2_common/boot_stage2/CMakeFiles/bs2_default.dir/build.make:96: pico-sdk/src/rp2_common/boot_stage2/bs2_default.elf] Error 1
[build] gmake[1]: *** [CMakeFiles/Makefile2:1885: pico-sdk/src/rp2_common/boot_stage2/CMakeFiles/bs2_default.dir/all] Error 2
```
I copied the SDK 1.5 into my install and still get build errors.  It is possible I need to do a fresh install as I have platformIO and other projects all banging around in my system now with numerous tweaks.  Unless those errors appear straight forward (not to me), maybe reload time is now.

**Reply by terjeio** [reply]
> I copied the SDK 1.5 into my install

Replacing or overwriting the current version? I replaced mine.

> Unless those errors appear straight forward

They are from the build system, not the grbrlHAL code. IMO best to reload.

**Reply by MadTooler** [reply]
`Replacing or overwriting the current version? I replaced mine.`
Replaced.  Renamed prior folder and inserted new.

**Comment by WheeSci** [comment]
I am having a problem figuring this out.  How is the SDK implemented?  I tried cloning from Github into the build folder but got an empty repo.  I downloaded the zipped repository and extracted the files into a directory (RP2040/pico-sdk) where my files for this are located.  When I tried to build the file, I get the error: SDK location not specified, Please set PICO_SDK_PATH or set PICO_SDK_FETCH_FROM_GIT.  Found these things in CMakeCache.txt. Got the path set, then got an error that toolchain path not specified.  That seems to need to be RP2040/pico-sdk/cmake/preload/.  Next error was in finding the compiler.  I downloaded and unzipped the tarball for gcc-arm-none-eabi but I'm getting the error message "gcc-arm-none-eabi-gcc" not found. This is a lot harder than I thought it would be. Is there any step by step instruction?

**Reply by terjeio** [reply]
There is an [installer](https://www.raspberrypi.com/news/raspberry-pi-pico-windows-installer/) for Windows that I used for my laptop.

**Comment by WheeSci** [comment]
Already had VSCode on my laptop. I updated it and installed the RPi Pico extension, loaded the blink example and tried to build. Got the same error message "cant open stdlib.h"

**Reply by terjeio** [reply]
The I suggest you [search the net](https://duckduckgo.com/?t=ffab&q=pi+pico+vscode+windows&ia=web) or ask on a [Pi forum](https://forums.raspberrypi.com/viewforum.php?f=143) for help - I do not have the required knowledge to support VSCode installations.

**Comment by WheeSci** [comment]
Thanks Terjeio, I appreciate what you do here.  This is just frustrating.  Anyway, for people who are facing the same issues, I will continue documenting my efforts.
It seems like the problem is that VSCode is not downloading and dealing with the SDK when I select the Pico extension.  On my Windows machine, I addressed this by going to the pico sdk github repository and downloaded as a zip file.  I created a folder: C://path/VSCode/include/ and unzipped the file in it.  The "path" term is specific to my machine. Having the sdk does not solve the problem however, you need to tell the compiler where to find the include files.  Here is a link to a [stack overflow thread ](https://stackoverflow.com/questions/37522462/visual-studio-code-include-stdio-h-saying-add-include-path-to-settings) that gives some info on this issue.  Scroll down to where you see a series of screen shots and follow the directions to add the path to where you have the include files.
Added the following:  C://path/VSCode/include/**
 Now when I try to build the project, I no longer get the error saying that the includes cannot be found. 
Now I have error messages saying 
  "Identifier PICO_OK is undefined"
  "Identifier true is undefined"
  "Identifier false is undefined"
Now I guess I have to find where these variables are defined.

**Reply by terjeio** [reply]
Recent clean installations of VSCode and the Pico extension has been fairly smooth for me - only problems with selecting Pico 2 variants for builds that somehow resolved itself.
